tsl/src/backup.c

```diff
@@ -0,0 +1,145 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+#include <postgres.h>
+#include <funcapi.h>
+#include <utils/builtins.h>
+#include <utils/pg_lsn.h>
+#include <access/xlog_internal.h>
+#include <access/xlog.h>
+#include <access/xact.h>
+#include <miscadmin.h>
+
+#include "errors.h"
+#include "dist_util.h"
+#include "remote/dist_commands.h"
+#include "backup.h"
+
+enum
+{
+	Anum_restore_point_node_name = 1,
+	Anum_restore_point_lsn,
+	_Anum_restore_point_max
+};
+
+static Datum
+create_restore_point_datum(TupleDesc tupdesc, const char *node_name, XLogRecPtr lsn)
+{
+	Datum values[_Anum_restore_point_max];
+	bool nulls[_Anum_restore_point_max] = { false };
+	HeapTuple tuple;
+
+	tupdesc = BlessTupleDesc(tupdesc);
+	values[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = CStringGetDatum(node_name);
+	if (node_name == NULL)
+		nulls[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = true;
+	values[AttrNumberGetAttrOffset(Anum_restore_point_lsn)] = LSNGetDatum(lsn);
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+	return HeapTupleGetDatum(tuple);
+}
+
+Datum
+create_distributed_restore_point(PG_FUNCTION_ARGS)
+{
+	const char *name = TextDatumGetCString(PG_GETARG_DATUM(0));
+	DistCmdResult *result_cmd;
+	FuncCallContext *funcctx;
+	XLogRecPtr lsn;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		MemoryContext oldctx;
+		TupleDesc tupdesc;
+		char *sql;
+
+		if (strlen(name) >= MAXFNAMELEN)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("restore point name is too long"),
+					 errhint("maximum %d characters", MAXFNAMELEN - 1)));
+
+		if (RecoveryInProgress())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errhint("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level is not sufficient for creating a restore point"),
+					 errhint(
+						 "wal_level must be set to \"replica\" or \"logical\" at server start.")));
+
+		if (!superuser())
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("must be superuser to create restore point")));
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("pg_create_restore_point() can be used to to create restore point "
+							 "on the data node.")));
+
+		/* create local restore point and on each data node */
+		lsn = XLogRestorePoint(name);
```

Comment:
So, I believe you have to explicitly lock the `remote_txn` table before creating the restore point, because we need to ensure that there aren't any concurrent distributed transactions that have committed and will become visible to us after the restore point.

Reply:
And, we need to ensure that other transactions hold their share lock until the end of their transactions.

Reply:
Also, perhaps it is better to take the local (AN) restore point after all the data node have taken restore points? While locking can ensure that it is kind of "atomic" I think it makes sense to have AN last. Thoughts?